src/Data/Nat/Primality/Factorisation.agda

```diff
@@ -0,0 +1,191 @@
+open import Data.List.Membership.Propositional.Properties using (∈-∃++)
+open import Data.List.Relation.Unary.All as All using (All; []; _∷_)
+open import Data.List.Relation.Unary.Any using (here; there)
+open import Data.List.Relation.Binary.Permutation.Propositional as ↭
+  using (_↭_; prep; swap; ↭-refl; refl; module PermutationReasoning)
+open import Data.List.Relation.Binary.Permutation.Propositional.Properties using (product-↭; All-resp-↭; shift)
+open import Data.Sum.Base using (inj₁; inj₂)
+open import Function.Base using (_$_; _∘_; _|>_; flip)
+open import Induction using (build)
+open import Induction.Lexicographic using (_⊗_; [_⊗_])
+open import Relation.Nullary.Decidable using (yes; no)
+open import Relation.Nullary.Negation using (contradiction)
+open import Relation.Binary.PropositionalEquality using (_≡_; refl; sym; trans; cong; module ≡-Reasoning)
+
+------------------------------------------------------------------------
+-- Core definition
+
+record PrimeFactorisation (n : ℕ) : Set where
+  field
+    factors : List ℕ
+    isFactorisation : product factors ≡ n
+    factorsPrime : All Prime factors
+
+open PrimeFactorisation public using (factors)
+open PrimeFactorisation
+
+------------------------------------------------------------------------
+-- Finding a factorisation
+
+-- this builds up three important things:
+-- * a proof that every number we've gotten to so far has increasingly higher
+--   possible least prime factor, so we don't have to repeat smaller factores
+--   over and over (this is the "m" and "rough" parameters)
+-- * a witness that this limit is getting closer to the number of interest, in a
+--   way that helps the termination checker (the "k" and "eq" parameters)
+-- * a proof that we can factorise any smaller number, which is useful when we
+--   encounter a factor, as we can then divide by that factor and continue from
+--   there without termination issues
+
+private
+  pattern 2+ n = suc (suc n)
+  pattern 2≤2+n = s≤s (s≤s z≤n)
+  pattern 1<2+n = 2≤2+n
+
+factorise : ∀ n → .{{NonZero n}} → PrimeFactorisation n
+factorise 1 = record
+  { factors = []
+  ; isFactorisation = refl
+  ; factorsPrime = []
+  }
+factorise n₀@(2+ _) = build [ <-recBuilder ⊗ <-recBuilder ] P facRec (n₀ , suc n₀ ∸ 4) 2-rough refl
+  where
+  P : ℕ × ℕ → Set
+  P (n , k) = ∀ {m} → .{{NonTrivial n}} → .{{NonTrivial m}} → m Rough n → suc n ∸ m * m ≡ k → PrimeFactorisation n
+
+  facRec : ∀ n×k → (<-Rec ⊗ <-Rec) P n×k → P n×k
+  -- Case 1: m * m > n, ∴ Prime n
+  facRec (n , zero) _ rough eq = record
+    { factors = n ∷ []
+    ; isFactorisation = *-identityʳ n
+    ; factorsPrime = rough∧>square⇒prime rough (m∸n≡0⇒m≤n eq) ∷ []
+    }
+  facRec (n@(2+ _) , suc k) (recFactor , recQuotient) {m@(2+ _)} rough eq with m ∣? n
+  -- Case 2: m ∤ n, try larger m, reducing k accordingly
+  ... | no m∤n = recFactor (≤-<-trans (m∸n≤m k (m + m)) (n<1+n k)) {suc m} (∤⇒rough-suc m∤n rough) $ begin
```

Comment:
Regarding the arguments to `recFactor`:
* The first argument can be simplified to: `s≤s (m∸n≤m k (2 * m))`; debatable whether it is an improvement, although shorter, because it exposes the `_<_` abtraction, where your proof-term respects it; better, I might think would be to have a lemma ` m∸n<m` in `Data.Nat.Properties` which has `m ∸ n < m` for `.{{NonZero n}}`?
* The second, visible-but-implicit, binding `{suc m}` can be omitted; again debatable whether it improves readability, but it is in any case governed by the appeal to `∤⇒rough-suc m∤n rough`, so the diligent reader could discover this for themselves...